sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala

```diff
@@ -46,10 +46,14 @@ class BulkWriter(container: CosmosAsyncContainer,
   // TODO: moderakh this requires tuning.
   // TODO: moderakh should we do a max on the max memory to ensure we don't run out of memory?
   private val cpuCount = SparkUtils.getNumberOfHostCPUCores
+  // each bulk writer allows up to maxPendingOperations being buffered
+  // there is one bulk writer per spark task/partition
+  // and default config will create one executor per core on the executor host
+  // so multiplying by cpuCount in the default config is too aggressive
   private val maxPendingOperations = writeConfig.bulkMaxPendingOperations
```

Comment:
what will happen is customer provide the maxPendingOperations?
The current guideline we give to customer is N *cores, are we also going to change the algorithm here? 
or add a new config so not to break change ?

Reply:
No - we leave that as is - (and should do so) - because whatever config customers who changed this manually have works for them
  private val maxPendingOperations = writeConfig.bulkMaxPendingOperations
    .getOrElse(DefaultMaxPendingOperationPerCore)